### PR TITLE
 Add interfaces for `Context` and `ContextKey`

### LINKED ISSUE #795: Add interfaces for `Context` and `ContextKey`
Simplifies mocking and allows using `enum`s as `ContextKey`.

### PR DESCRIPTION
**Pull request by Nevay**
Resolves #795.

### CODE REVIEW

File: src/API/Trace/SpanBuilderInterface.php [hunk 1]
```diff
@@ -15,14 +15,14 @@
...
-     * Sets the parent {@see Context} to use.
+     * Sets the parent {@see ContextInterface} to use.
      *
      * If no {@see SpanInterface} is available in the provided context, the resulting span will become a root span,
      * as if {@see SpanBuilderInterface::setNoParent} was called.
      *
      * Defaults to {@see Context::getCurrent} when {@see SpanBuilderInterface::startSpan} was called if not explicitly set.
      */
-    public function setParent(Context $parentContext): SpanBuilderInterface;
+    public function setParent(ContextInterface $parentContext): SpanBuilderInterface;
```
Comment (Nevay) [inline src/API/Trace/SpanBuilderInterface.php:25]:
```suggestion
    public function setParent(ContextInterface|false|null $parentContext): SpanBuilderInterface;
```
Should this be changed (same typehint is already used in metrics API/SDK for context parameters) and `::setNoParent()` be removed?

File: src/Context/ImplicitContextKeyedInterface.php [hunk 2]
```diff
@@ -26,8 +26,8 @@ interface ImplicitContextKeyedInterface
     public function activate(): ScopeInterface;
 
     /**
-     * Returns a new {@see Context} created by setting `$this` into the provided [@see Context}.
-     * It is suggested to use {@see Context::withContextValue()} instead of using this method directly.
+     * Returns a new {@see ContextInterface} created by setting `$this` into the provided [@see ContextInterface}.
+     * It is suggested to use {@see ContextInterface::withContextValue()} instead of using this method directly.
```
Comment (Nevay) [inline src/Context/ImplicitContextKeyedInterface.php:30]: What is the reason for this suggestion? Using this method feels more fluent.
```php
$spanBuilder
    ->setParent($parent)
    ->startSpan()
    ->storeInContext($parent)
    ->activate()
```

File: src/Context/Propagation/NullPropagator.php [hunk 3]
```diff
@@ -5,6 +5,7 @@
 namespace OpenTelemetry\Context\Propagation;
 
 use OpenTelemetry\Context\Context;
+use OpenTelemetry\Context\ContextInterface;
 
 class NullPropagator implements TextMapPropagatorInterface
```
Comment (Nevay) [inline src/Context/Propagation/NullPropagator.php:10]: Should be removed, duplicates `NoopTextMapPropagator`.

File: src/API/Trace/SpanBuilderInterface.php (hunk 1 above)
Comment (brettmc) [reply src/API/Trace/SpanBuilderInterface.php:25]: Yes that makes sense. Are false and null equivalent, here?

File: src/Context/Propagation/NullPropagator.php (hunk 3 above)
Comment (brettmc) [reply src/Context/Propagation/NullPropagator.php:10]: agree

File: src/Context/ImplicitContextKeyedInterface.php (hunk 2 above)
Comment (brettmc) [reply src/Context/ImplicitContextKeyedInterface.php:30]: It's before my time, I think it's ok to remove the comment.

File: src/SDK/Trace/SpanBuilder.php [hunk 4]
```diff
@@ -51,17 +53,9 @@ public function __construct(
     }
 
     /** @inheritDoc */
-    public function setParent(Context $parentContext): API\SpanBuilderInterface
+    public function setParent($context): API\SpanBuilderInterface
```
Comment (brettmc) [inline src/SDK/Trace/SpanBuilder.php:56]: I think we should document here (or in the API?) what ContextInterface/null/false mean

Comment (Nevay) [reply src/SDK/Trace/SpanBuilder.php:56]: `SpanBuilderInterface::setParent()`:
```
 * @param ContextInterface|false|null $context the parent context, null to use the
 *        current context, false to set no parent
```

**Review by brettmc — APPROVED** [review]
approved, some conflicts to be resolved before we can merge

**Pull request merged**